dask/dataframe/utils.py

```diff
@@ -382,7 +382,13 @@ def meta_nonempty_object(x):
 @meta_nonempty.register(pd.DataFrame)
 def meta_nonempty_dataframe(x):
     idx = meta_nonempty(x.index)
-    data = {i: _nonempty_series(x.iloc[:, i], idx=idx) for i, c in enumerate(x.columns)}
+    if len(x.dtypes.unique()) == 1:
```

Comment:
This seems like it would help in the case where we have a single dtype, but would not help if we have many more columns than dtypes. 

Perhaps we could generalize this to create a nonempty series for every dtype, and then copy that series for however many columns there are of that dtype. 

Also, I wonder if there are dtypes which need special consideration, like categoricals (I don't know that this is the case, I'm just curious)

Reply:
I was playing around with

```
dts = x.dtypes.unique()
data = dict()
for dt in dts:
    dt_col_idx = np.where(x.dtypes == dt)[0]
    s = _nonempty_series(x.iloc[:, dt_col_idx[0]], idx=idx)
    update_map = zip(dt_col_idx, [s] * len(dt_col_idx))
    data.update(update_map)
```

which is much faster than the pushed implementation when it works. But, as you guessed @mrocklin , there are problems with categoricals. Namely, `np.where` doesn't work with them. I also played around with `df.select_dtypes`, but that was as slow as the original implementation. Any suggestions on how to get around the `np.where` problem?

Reply:
Also, `x.dtypes.unique()` treats each categorical as unique

Reply:
Maybe we do this efficiency trick for some set of types that are known to be sensible, and fall back to the old/inefficient way of doing things for everything else.  

For example, I wonder if all numpy dtypes are safe

```python
In [5]: import pandas as pd
df =
In [6]: df = pd.DataFrame({"x": [1, 2, 3], "y": ["a", "b", "a"]})

In [7]: df["y"] = df.y.astype("category")

In [8]: df.dtypes
Out[8]:
x       int64
y    category
dtype: object

In [9]: isinstance(df.dtypes["x"], np.dtype)
Out[9]: True

In [10]: isinstance(df.dtypes["y"], np.dtype)
Out[10]: False
```

Reply:
Sorry, I haven't followed this too closely, but pandas does require that dtypes are hashable. Not sure if that's relevant.

> Also, x.dtypes.unique() treats each categorical as unique

This is the correct behavior, right? I would expect that categoricals with different categories get treated differently. Also things like `PeriodDtype(freq="D")` vs. `PeriodDtype(freq="H")` would be distinct.